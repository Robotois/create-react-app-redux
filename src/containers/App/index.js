import React from 'react';
import { Route } from 'react-router-dom';

import Pokemons from '../Pokemons';
import Pokemon from '../Pokemons/item';

export default class App extends React.Component {
  render() {
    return (
      <div>
        <main>
          <Route exact path="/" component={Pokemons} />
          <Route exact path="/pokemon/:id" component={Pokemon} />
        </main>
      </div>
    );
  }
}
