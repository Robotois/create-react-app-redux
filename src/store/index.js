import { createStore as reduxCreateStore, applyMiddleware, compose, combineReducers } from 'redux';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import thunk from 'redux-thunk';
import createHistory from 'history/createBrowserHistory';

import clientMiddleware from '../utils/clientMiddleware';
import ApiClient from '../utils/ApiClient';
import appReducer, { rootReducer } from '../modules/rootReducer';

export const history = createHistory();
const enhancers = [];

if (process.env.NODE_ENV === 'development') {
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__;

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension());
  }
}

const createStore = client =>
  reduxCreateStore(
    connectRouter(history)(rootReducer(combineReducers(appReducer))),
    compose(
      applyMiddleware(clientMiddleware(client), routerMiddleware(history), thunk),
      ...enhancers,
    ),
  );

const store = createStore(new ApiClient());

/**
 * Register async loaded reducers in store and replace
 * current state-reducer with the a new reducer
 *
 * @export
 * @param {Object} store - the store object
 * @param {Object} asyncReducer - async reducer modules
 */
store.asyncReducers = {};

function replaceReducers(defaultReducers) {
  const merged = Object.assign({}, defaultReducers, store.asyncReducers);
  const combined = rootReducer(combineReducers(merged));
  store.replaceReducer(combined);
}

export function injectAsyncReducers(asyncReducers) {
  const injectReducers = Object.keys(asyncReducers).reduce((toAdd, reducer) => {
    // do not replace already existing reducer unless it was (potentially) hot-reloaded
    if (store.asyncReducers[reducer] && !module.hot) {
      delete toAdd[reducer];
    }

    return toAdd;
  }, asyncReducers);

  store.asyncReducers = Object.assign({}, store.asyncReducers, injectReducers);
  replaceReducers(appReducer);
}

/* istanbul ignore next */
if (module.hot) {
  module.hot.accept('../modules/rootReducer', () => {
    const nextReducer = require('../modules/rootReducer').default; // eslint-disable-line global-require

    store.replaceReducer(nextReducer);
  });
}

export default store;
