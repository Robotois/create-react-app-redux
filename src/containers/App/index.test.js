import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../store';
import { MemoryRouter } from 'react-router-dom';

import App from './';

const Wrapper = () => (
  <Provider store={store}>
    <App />
  </Provider>
);

it('renders without crashing', () => {
  const component = renderer.create(
    <MemoryRouter>
      <Wrapper />
    </MemoryRouter>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
