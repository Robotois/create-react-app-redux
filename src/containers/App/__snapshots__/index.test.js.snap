// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <header>
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    Â 
    <a
      href="/about-us"
      onClick={[Function]}
    >
      About
    </a>
  </header>
  <main>
    <div>
      <h1>
        Home
      </h1>
      <p>
        Count: 
        0
      </p>
      <p>
        <button
          onClick={[Function]}
        >
          Increment
        </button>
        <button
          disabled={false}
          onClick={[Function]}
        >
          Increment Async
        </button>
      </p>
      <p>
        <button
          onClick={[Function]}
        >
          Decrement
        </button>
        <button
          disabled={false}
          onClick={[Function]}
        >
          Decrement Async
        </button>
      </p>
      <p>
        <button
          onClick={[Function]}
        >
          Go to about page via redux
        </button>
      </p>
    </div>
  </main>
</div>
`;
